pango/src/attribute.rs

```diff
@@ -217,4 +220,199 @@ impl Attribute {
             (*stash.0).end_index = index;
         }
     }
+    pub fn downcast<T: IsAttribute>(self) -> Result<T, Attribute> {
+        unsafe {
+            if T::ATTR_TYPES.contains(&self.attr_class().type_()) {
+                Ok(from_glib_full(self.to_glib_full()))
+            } else {
+                Err(self)
+            }
+        }
+    }
+
+    pub fn downcast_ref<T: IsAttribute>(&self) -> Option<&T> {
+        unsafe {
+            if T::ATTR_TYPES.contains(&self.attr_class().type_()) {
+                Some(&*(self as *const Attribute as *const T))
+            } else {
+                None
+            }
+        }
+    }
+}
+
+pub unsafe trait IsAttribute:
+    FromGlibPtrFull<*const ffi::PangoAttribute>
+    + FromGlibPtrFull<*mut ffi::PangoAttribute>
+    + std::convert::AsRef<crate::Attribute>
+    + 'static
+{
+    const ATTR_TYPES: &'static [AttrType];
+    fn upcast(self) -> Attribute;
+    fn upcast_ref(&self) -> &Attribute;
+}
+
+macro_rules! define_attribute_struct {
+    ($rust_type:ident, $ffi_type:path, $attr_types:expr) => {
+
+        #[cfg(any(feature = "v1_44", feature = "dox"))]
+        #[cfg_attr(feature = "dox", doc(cfg(feature = "v1_44")))]
+        glib::wrapper! {
+            #[derive(Debug)]
+            pub struct $rust_type(Boxed<$ffi_type>);
+
+            match fn {
+                copy => |ptr| ffi::pango_attribute_copy(ptr as *const ffi::PangoAttribute) as *mut $ffi_type,
+                free => |ptr| ffi::pango_attribute_destroy(ptr as *mut ffi::PangoAttribute),
+                type_ => || ffi::pango_attribute_get_type(),
+            }
+        }
+
+        #[cfg(not(any(feature = "v1_44", feature = "dox")))]
+        glib::wrapper! {
+            #[derive(Debug)]
+            pub struct $rust_type(Boxed<$ffi_type>);
+
+            match fn {
+                copy => |ptr| ffi::pango_attribute_copy(ptr as *const ffi::PangoAttribute) as *mut $ffi_type,
+                free => |ptr| ffi::pango_attribute_destroy(ptr as *mut ffi::PangoAttribute),
+            }
+        }
+
+        impl $rust_type {
+            #[doc(alias = "pango_attribute_equal")]
+            fn equal(&self, attr2: &$rust_type) -> bool {
```

Comment:
```suggestion
            fn equal<T: IsAttribute>(&self, attr2: &T) -> bool {
```

Should also work or not? More generic